src/com/rabbitmq/client/impl/recovery/AutorecoveringConnection.java

```diff
@@ -428,18 +470,43 @@ private void recoverBlockedListeners() {
         }
     }
 
-    private void recoverConnection() throws IOException, InterruptedException {
-        boolean recovering = true;
-        while (recovering) {
+	//Returns true if the connection was recovered, 
+	//false if application initiated shutdown while attempting recovery.  
+    private boolean recoverConnection() throws IOException, InterruptedException {
+        while (!manuallyClosed) //manuallyClosed marked volatile -ensures thread cannot cache value of manuallyClosed.
+		{
             try {
-                this.delegate = this.cf.newConnection();
-                recovering = false;
+				RecoveryAwareAMQConnection newConn = this.cf.newConnection();
+				synchronized(recoveryLock)
+				{
+					if (!manuallyClosed)
+					{
+						//This is the standard case.
+						this.delegate = newConn;					
+						return true;
+					}
+				}
+
+				//This is the once in a blue moon case.  
+				//Application code just called close as the connection
+				//was being re-established.  So we attempt to close the newly created connection.
+				try
+				{
+					newConn.close(); //Is there a standard error code/shutdown message to use here? 
+				}
+				catch (Exception e)
+				{
+					//Ugh, we just leaked the new connection object, hopefully some kind of heartbeat/expiration timeout will take care of it.
```

Comment:
If the connection is kept alive, there is no expiration and heartbeats, in fact, will *keep it open* (and the object will have references). Please do not leave comments you are not 100% sure about, as incorrect comments
are worse than not having any.

Use `Connection#abort` as recommended above.